examples/run-blockchain/index.ts

```diff
@@ -0,0 +1,118 @@
+import VM from '../../'
+
+import Account from 'ethereumjs-account'
+import * as utils from 'ethereumjs-util'
+import { promisify } from 'util'
+
+const Trie = require('merkle-patricia-tree/secure')
+const Block = require('ethereumjs-block')
+const Blockchain = require('ethereumjs-blockchain')
+const BlockHeader = require('ethereumjs-block/header.js')
+const level = require('level')
+const levelMem = require('level-mem')
+const testData = require('./test-data')
+
+const rlp = utils.rlp
+
+async function main() {
+  const state = new Trie()
+  const blockchainDB = levelMem()
+
+  const hardfork = testData.network.toLowerCase()
+
+  const blockchain = new Blockchain({ db: blockchainDB, hardfork })
+  blockchain.ethash.cacheDB = level('./.cachedb')
+
+  const vm = new VM({
+    state: state,
+    blockchain: blockchain,
+    hardfork,
+  })
+
+  vm.on('beforeTx', (tx: any) => {
+    tx._homestead = true
+  })
+
+  vm.on('beforeBlock', (block: any) => {
+    block.header.isHomestead = function() {
+      return true
+    }
+  })
+
+  await setupPreConditions(state, testData)
+
+  await setGenesisBlock(blockchain, hardfork)
+
+  await putBlocks(blockchain, hardfork, testData)
+
+  await vm.runBlockchain(blockchain)
+
+  const blockchainHead = await promisify(vm.blockchain.getHead.bind(vm.blockchain))()
+  state.root = blockchainHead.header.stateRoot
+
+  console.log('--- Finished processing the BlockChain ---')
+  console.log('New head:', '0x' + blockchain.meta.rawHead.toString('hex'))
+  console.log('Expected:', testData.lastblockhash)
+}
+
+async function setupPreConditions(state: any, testData: any) {
+  for (const address of Object.keys(testData.pre)) {
+    const acctData = testData.pre[address]
+    const account = new Account()
+
+    account.nonce = utils.toBuffer(acctData.nonce)
+    account.balance = utils.toBuffer(acctData.balance)
+
+    const storageTrie = state.copy()
```

Comment:
Shouldn't this be an empty trie? Why a copy?

Reply:
Ah I was trying to understand this recently myself:

The goal here has been to use the same underlying leveldb instance to store both the accounts and their storage tries. So this `copy` is creating a new trie with reference to the same db. When you'd want to query a storageTrie, you could copy the state trie, set its root to the storageRoot of that account, and `get`, `put`, etc values.

Reply:
I see! 

Shouldn't it be the same than the blockchain's db? I updated this, passing the db explicitly.

Reply:
Doesn't necessarily need to be the same db with blockchain. I suppose it could though.
I don't also see a reason why it has to be the same db between state and storage tries. I think it was just easier to use one db.

Reply:
I'm not sure where the code should be stored. The Acount only has a hash of it. Should it be in the state db? Or the account's storage db?

I want to make this example as simple as possible :)

Reply:
I don't think it matters, as long as the same db is later on queried to get the code. `await setCode(account, state, codeBuf)` is currently storing it in the state db (where storage tries are also stored). `Account.getCode` and `Account.setCode` are used to store and fetch the code.
I imagine this can be confusing. Not sure if using the same db is clearer or different ones. Maybe some comments could help clarify

Reply:
Ohh, got it! I think what was *really* confusing me was that `Account#setCode` and `Account#getCode` take a `Trie` just to use their `db` via `putRaw`/`getRaw`. I thought they were modifying that `Trie`, which they don't.

Reply:
Exactly. I think there's room for improvement around the design of this stuff :)